R/convert_output_types.R

```diff
@@ -0,0 +1,157 @@
+#' terminal output types, and the resulting output will only contain the terminal
+#' output types. See details for supported conversions.
+#'
+#' @param model_out_tbl an object of class `model_out_tbl` containing predictions
+#'    with a single, unique value in the `output_type` column.
+#' @param terminal_output_type_id a named list indicating the desired output types and
+#'   associated output type IDs. List item name and value pairs may be as follows:
+#'   - `mean`: `NA` (no associated output type ID)
+#'   - `median`: `NA` (no associated output type ID)
+#'   - `quantile`: a numeric vector of probability levels
+#'
+#' @details Currently, only `"sample"` can be converted to `"mean"`, `"median"`, or `"quantile"`
+#'
+#' @examples
+#' # We illustrate the conversion between output types using normal distributions
+#' ex_quantiles <- c(0.25, 0.5, 0.75)
+#' model_out_tbl <- expand.grid(
+#'   stringsAsFactors = FALSE,
+#'   group1 = c(1,2),
+#'   model_id = "A",
+#'   output_type = "sample",
+#'   output_type_id = 1:100
+#' ) |>
+#' dplyr::mutate(value = rnorm(200, mean = group1))
+#'
+#' # Multiple output type conversions in one call
+#' convert_output_type(model_out_tbl,
+#'   terminal_output_type_id = list("quantile" = ex_quantiles, "median" = NA))
+#'
+#' # Single output type conversion
+#' convert_output_type(model_out_tbl, terminal_output_type_id = list("quantile" = ex_quantiles))
+#'
+#' @return object of class `model_out_tbl` containing (only) predictions of the
+#'   terminal output_type(s) for each unique combination of task IDs for each model
+#'
+#' @export
+#' @importFrom rlang .data
+convert_output_type <- function(model_out_tbl, terminal_output_type_id) {
+  # validations
+  terminal_output_type <- names(terminal_output_type_id)
+  model_out_tbl <- validate_conversion_inputs(model_out_tbl, terminal_output_type, terminal_output_type_id)
+
+
+  terminal_output_type |>
+    purrr::map(convert_single_output_type, terminal_output_type_id, model_out_tbl) |>
+    purrr::list_rbind() |>
+    as_model_out_tbl()
+}
+
+#' Convert single output type
+#' @noRd
+#' @importFrom rlang .data
+convert_single_output_type <- function(terminal_output_type, terminal_output_type_id, model_out_tbl) {
+  task_id_cols <- subset_task_id_names(names(model_out_tbl))
+  grouped_model_outputs <- dplyr::group_by(model_out_tbl, dplyr::across(dplyr::all_of(c("model_id", task_id_cols))))
+  terminal_output_type_id <- terminal_output_type_id[[terminal_output_type]]
+
+  if (terminal_output_type == "mean") {
+    converted_outputs_tmp <- dplyr::reframe(
+      grouped_model_outputs,
+      output_type_id = terminal_output_type_id,
+      value = mean(.data[["value"]])
+    )
+  } else if (terminal_output_type == "median") {
+    converted_outputs_tmp <- dplyr::reframe(
+      grouped_model_outputs,
+      output_type_id = terminal_output_type_id,
+      value = stats::median(.data[["value"]])
+    )
```

Comment:
The `mean` and `median` as `NULL` would work if you did this:

```suggestion
  if (terminal_output_type == "mean") {
    converted_outputs_tmp <- dplyr::reframe(
      grouped_model_outputs,
      output_type_id = NA,
      value = mean(.data[["value"]])
    )
  } else if (terminal_output_type == "median") {
    converted_outputs_tmp <- dplyr::reframe(
      grouped_model_outputs,
      output_type_id = NA,
      value = stats::median(.data[["value"]])
    )
```

Reply:
Note that this is for demonstration only, not intended to be a suggestion unless you change the function signature